### PR TITLE
Add support for GPTQ Marlin

### PR DESCRIPTION
**Pull request by danieldk**
# What does this PR do?

Add support for GPTQ Marlin kernels. GPTQ Marlin extends the Marlin kernels to support common GPTQ configurations:

- bits: 4 or 8
- groupsize: -1, 32, 64, or 128
- desc_act: true/false

Using the GPTQ Marlin kernels requires repacking the parameters in the Marlin quantizer format.

The kernels were contributed by Neural Magic to VLLM. We vendor them here for convenience.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: server/text_generation_server/layers/marlin.py [hunk 1]
```diff
@@ -1,23 +1,220 @@
...
+GPTQ_MARLIN_BITS = [4, 8]
+GPTQ_MARLIN_GROUP_SIZES = [-1, 32, 64, 128]
 MARLIN_TILE_SIZE = 16
 
 
+def _check_marlin_kernels():
+    if not has_sm_8_0:
+        raise NotImplementedError(
+            "Using quantized marlin models requires CUDA capability 8.0 or later"
+        )
```
Comment (fxmarty) [inline server/text_generation_server/layers/marlin.py:30]: Can you add a check on `SYSTEM == "cuda"` here?

File: server/text_generation_server/utils/weights.py [hunk 2]
```diff
@@ -9,6 +10,15 @@
 from text_generation_server.utils.log import log_once
 
 
+@dataclass
+class _GPTQParams:
```
Comment (drbh) [inline server/text_generation_server/utils/weights.py:14]: not important: but is there a reason this class is prefixed with `_`?

**Review by drbh — APPROVED** [review]
lgtm!

**Review by fxmarty — APPROVED** [review]

File: server/text_generation_server/utils/weights.py (hunk 2 above)
Comment (danieldk) [reply server/text_generation_server/utils/weights.py:14]: To make it extra clear that the class is not meant for external consumption (not a public API).

**Pull request merged**